HPCC-22136 Import WU from ZAP report

wangkx opened the pull request:
1. Upgrade the _createWorkUnit() to support wuXMLPTree.
2. Add importWorkUnit() to IWorkUnitFactory:
   Unzip a ZAP report into unique esp data folder;
   Update WU XML;
   Calls the  _createWorkUnit() with the WU XML.
3. Add WORKUNIT_API importWorkunitFromZAPFile() which calls the
importWorkUnit().
4. Add WsWorkunit code to accept uploaded ZAP report and call
importWorkunitFromZAPFile().
5. Update WsWuInfo::getWorkunitThorLogInfo() to get Thor Log
Information from ZAP report.
6. Add the code to read log files (eclagent log and thor logs)
from ZAP files.

Also:
1. Store ESP name when WsWorkunits service is loaded. The name
is used to find out the esp data folder from getConfigurationDirectory().
2. Modify CHttpRequest::createUploadFile: If no netAddress is specified
for the upload target, create a local file as the upload target.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

wangkx committed:
HPCC-22136 Import WU from ZAP report

1. Upgrade the _createWorkUnit() to support wuXMLPTree.
2. Add importWorkUnit() to IWorkUnitFactory:
   Unzip a ZAP report into unique esp data folder;
   Update WU XML;
   Calls the  _createWorkUnit() with the WU XML.
3. Add WORKUNIT_API importWorkunitFromZAPFile() which calls the
importWorkUnit().
4. Add WsWorkunit code to accept uploaded ZAP report and call
importWorkunitFromZAPFile().
5. Update WsWuInfo::getWorkunitThorLogInfo() to get Thor Log
Information from ZAP report.
6. Add the code to read log files (eclagent log and thor logs)
from ZAP files.

Also:
1. Store ESP name when WsWorkunits service is loaded. The name
is used to find out the esp data folder from getConfigurationDirectory().
2. Modify CHttpRequest::createUploadFile: If no netAddress is specified
for the upload target, create a local file as the upload target.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith please review

wangkx committed:
HPCC-22136 Revised based on review

1. Rolled back the changes for _createWorkUnit();
2. Added IWorkunit.import();
3. Used new WUID as upzip folder name;
4. Revised the CHttpRequest::createUploadFile();
5. Renamed the getWorkunitLogFromZAPFile() to getWorkunitLogSingleFile();
6. Renamed the getWorkunitThorSlaveLogFromZAPFile() to getWorkunitThorSlaveLogSingleFile();
7. Added a 'return;';
8. Added updateSecondaries() to import WU into Cassandra WU search tables;
9. Added getThorGroup() and setThorGroup().

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -5389,15 +5556,44 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
+            const char *wuidUnique = root->queryName();
+            wuXMLPTree->renameProp("/", wuidUnique);
+
+            Owned<IPropertyTree> progressPTree = pruneBranch(wuXMLPTree, "GraphProgress[1]");
```
The [1] shouldn't be necessary, it implies there could be >1 GraphProgress node..

jakesmith commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5062,171 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+            //update QueryAssociatedFiles in WU XML;
+            updateWUQueryAssociatedFilesAttrs(localIP);
+
+            readGraphProgressFileToWUPTree();
```
it's a little inefficient to load the file and set it into the wuTree.
For it later to be removed (pruned) then added separately to /GraphProgress, when really it was separate in the 1st place.

I think this would better as a separate step. (but still within CWorkUnitFactory::importWorkUnit) after the creation of the wuid and import of it's xml (see comments above), then add progress tree.

e.g. something like:
```
    CImportWorkUnitHelper helper(zapReportFileName, user);
    Owned<IWorkUnit> newWuid = createWorkUnit(app, user, secMgr, secUser);
    helper.setUNZIPDir(directories, component, instance, newWuid->queryWuid());
    helper.unzipZAPReport(zapReportFilePath, zapReportPassword);   
    helper.buildWUPTreesFromZAPReport(); // keeps progress tree separately

    Owned<IPropertyTree> wuidTree = helper.getWorkUnitTree();
    newWuid->import(wuidTree);
    Owned<IPropertyTree> progressTree = helper.getPropgressTree();
    importProgress(newWuid->queryWuid(), progressTree);
```

wangkx replied:
@jakesmith In fact, I DO like this approach better because it is much easy to understand and to implement. The existing approach is based on your comment: 'I would be better/more efficient to do it in one call. I would add a new createWorkUnit method that takes a IPropertyTree. Load the archive workunit xml to a ptree, patch up where necessary, than add via the new createWorkUnit, that copies the content into the newly created uniquely named wuid.' As your comment said, the existing approach has only one call to Dali WUs Tree. Is the existing approach more efficient than this approach?

jakesmith replied:
@wangkx - efficiency is important, whilst not a critical area, it could be time consuming and is easy to avoid adding then removing the progress tree here.
a new createWorkUnit method (as opposed to altering the existing) would have worked too, if you had passed the workunit tree and progress tree separately.
The issue here is you add it to a loaded workunit xml , but then prune it later, before making the new connection for the GraphProgress.

The both new and approach, must make 2 connections - one for the /WorkUnit/... and one for the GraphProgress, there is no difference in efficiency wrt as far as that is concerned, in both cases it has to copy 2 trees. So as far a Dali traffic is concerned there is no difference.
It's the local copying/pruning of the GraphProgress that can be avoided with this approach.

wangkx replied:
ok

jakesmith commented on esp/bindings/http/platform/httptransport.cpp:
```diff
@@ -1971,6 +1971,10 @@ IFile* CHttpRequest::createUploadFile(const char * netAddress, const char* fileP
 
+    //If no netAddress is specified, create a local file.
+    if (isEmptyString(netAddress))
+        return createIFile(tmpFileName);
```
Not new, but looking at the code above this..
It looks rather odd, reversing twice, strchr and writing inside StringBuffer..
What is it trying to do exactly?

it's called tmpFileName, should it be using createUniqueFile instead?

wangkx replied:
The code may be replaced by pathTail(). Will do.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2051,6 +2145,20 @@ void WsWuInfo::getWorkunitThorSlaveLog(IPropertyTree* directories, const char *p
+    if (!logfile)
+        throw MakeStringException(ECLWATCH_CANNOT_OPEN_FILE, "Cannot open %s.", logFileName.str());
+
+    getWorkunitLogFromZAPFile(logfile, logFileName, buf, outFile);
```
I haven't looked at changes here and above very carefully yet,
but we do plan to have logs per workunit.
So it would better if possible, there was little or no mention of "ZAP" in related changed of code that is looking at workunits whether they are 'original' or imported.

It should if at all possible be agnostic and the code look as similar as possible.

As I say, I haven't fully looked at all the changes involved here,
and perhaps there are some cases where the workunit has references to clusters (for example) that don't exist in the environment they've been imported into, and the current code relies on them to some extent.

Let's have an offline discussion re. viewing/using imported workunits, vs existing, to explore what differences are needed..

jakesmith commented:
@wangkx - I've given this a 1st pass, and made various comments.
In general it looks good, but I have some questions, and I think it could be restructured a little.

I did not review the code relating to viewing imported files (that refers to them a ZAP workunits), I got the feeling that there was too much special casing, which I would hope there wouldn't.
Perhaps necessary for backward compatibility, if old code relies on lots of lookups into (for example) clusters in the environment.. I'm not sure.
Let's discuss that particular aspect offline.

Did you test the Cassandra implementation?

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1975,7 +2038,32 @@ void WsWuInfo::getWorkunitThorLog(const char* fileName, MemoryBuffer& buf, const
+        if (!oFile)
+            throw MakeStringException(ECLWATCH_CANNOT_OPEN_FILE, "Cannot open %s.", outFile);
+
+        copyFile(oFile, iFile);
```
Currently getWorkunitLogFromZAPFile is both writing to the file and copying to buffer.
There's a:
```
return;
```
missing here

wangkx replied:
Will add back.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
+        else
+        {
+            numberOfSlaveLogs = getNumberOfSlavesByClusterName();
+            thorMasterLogDateSearchString.set("/thormaster.");
```
can u always deduce # of slaves from the log (names) if number of slaves is required?
What if slave logs aren't available, because zap report didn't contain them?

What is numberOfSlaveLogs needed for?
I see the:
```
                thorLog->setNumberSlaves(numberOfSlaveLogs);
```
..below. Is that so Eclwatch can go to a particular slave log?

If so, could deduce # of slaves from slave log names if available in imported helpers.
Just a thought, perhaps workunit should always, contain group and pattern to slave logs, so that doesn't have to go lookup stuff in environment which may be out of date anyway.
Then an imported workunit would not be particularly different, would also contain group and pattern in same way.

jakesmith replied:
There's:
```
1142:            StringBuffer groupName;
1143:            getClusterThorGroupName(groupName, processName.str());
```

that will likely not make much sense for an imported version.
i.e. the processName will not exist in the local environment.

what is groupName needed for?

wangkx replied:
> .below. Is that so Eclwatch can go to a particular slave log?
Yes, ECLWatch has to display a number which a user may choose to download a slave log.  
> what is groupName needed for?
The groupName is used for standard (not imported) WU to get slave log.

jakesmith replied:
So what if instead there were methods in the workunit like:
```
IGroup *IConstWorkUnit::getThorGroup();
```
and:
```
const char *IConstWorkUnit::getSlaveLogPattern();
```

and Thor could published the group and the slave log pattern to the workunit.
For legacy workunits (that didn't have published group/pattern), it could as you do, get the group based on getClusterThorGroupName, and the slave pattern from the environment directories.

Then this code could be agnostic to where workunit came from

wangkx replied:
Here, I need: ThorGroup, # of slaves, and the dates of the thor logs. The getThorGroup() returns the ThorGroup. What will getSlaveLogPattern() return?

wangkx replied:
If we have the logPerWorkunit, I still need to know the number of slaves (for example, 50) so that ECLwatch can display a list of numbers (from 1 to 50) for a user to download a specific slave log. So, Should the pattern be: /var/log/HPCCSystems/mythor/thorslave.*_of_50.2019_11_1[2-3].log or /var/log/HPCCSystems/mythor/W20191113-140205/thorslave.*_of_50.log ?

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
-            IWARNLOG("Cannot find TargetClusterInfo for workunit %s", cw->queryWuid());
-            return countThorLog;
+            numberOfSlaveLogs = getNumberOfSlavesFromZAPFiles();
+            thorMasterLogDateSearchString.set("_thormaster.");
```
Isn't there a directory per workunit?
If so, can't pattern be the same, i.e. "thormaster." ?

wangkx replied:
When a ZAP report is created, the '_' is added in the front of the 'thormaster.'

jakesmith replied:
ok, you _could_ when you unzip it, normalize it so filenames are same..
I think a directory per workunit with consistently named logs is what we want.
Not sure why ZAP file isn't like that too, any reason it couldn't be?

wangkx replied:
When creating a ZAP report, all of WU files are retrieved/created into one folder with one layer. A WU may have thor logs from >1 thor instances. To identify them, thor instance name is used as a prefix of a thor master log, like: analyt_thor400_90_a_thormaster.2019_02_14

jakesmith replied:
right ok, yes that makes sense.
In that case could use a regular pattern in both cases "*thormaster*.log" or similar..
To reduce amount of conditional code?

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
-        if (!clusterInfo)
+        unsigned numberOfSlaveLogs = 0;
+        StringAttr thorMasterLogDateSearchString;
+        if (debugImportedStr.length()) //WU imported from a ZAP report
```
As discussed offline, would be good to remove references to 'improted' and ZAP where possible in new code, and instead distinguish where necessary as 'logsPerWorkUnit' flag or similar.
Since we plan to have workunits per workunit in general anyway.

wangkx replied:
Should the coming import() function insert a 'logsPerWorkUnit' flag into the WU tree if not exist?

jakesmith replied:
for now I think the 'imported' string you added in the publish info is fine.
But internally refer to that once, and use a flag everywhere for conditional code where necessary.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1122,7 +1178,7 @@ unsigned WsWuInfo::getWorkunitThorLogInfo(IArrayOf<IEspECLHelpFile>& helpers, IE
 
                 const char* pStr = logName.str();
-                const char* ppStr = strstr(pStr, "/thormaster.");
+                const char* ppStr = strstr(pStr, thorMasterLogDateSearchString.get());
```
why is this necessary?
To get these logs, it has called cw->getLogs("Thor", ...)
so they explicitly thor log files already (no matter what name they have)

jakesmith replied:
Also note:
```
ppStr += 12
```
would be wrong if thorMasterLogDateSearchString varies. Should avoid constants like this

wangkx replied:
The code here is to find out the log date. Some WU may run for > 1 days. ECLWatch has to display a dialog box for a user to choose a thor slave using the log date, as well as thor process name and a slave ID.

jakesmith replied:
But should not use constants like '12'.
Use a const string and strlen if necessary for readability/maintainability.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1896,6 +1952,13 @@ void WsWuInfo::getWorkunitEclAgentLog(const char* fileName, const char* agentPid
+
+    if (debugImportedStr.length()) //WU imported from a ZAP report
+    {
+        getWorkunitLogFromZAPFile(rFile, fileName, buf, outFile);
```
can we switch from debugImportedStr to a flag like e.g. logsPerWorkunit,
and from getWorkunitLogFromZAPFile and similar to getWorkunitLogSingleFile

wangkx replied:
We may switch from debugImportedStr to a flag like e.g. logsPerWorkunit if we do not need the flag to find out the log date and # of slaves from ZAP.

wangkx replied:
Will do

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2051,6 +2145,20 @@ void WsWuInfo::getWorkunitThorSlaveLog(IPropertyTree* directories, const char *p
+void WsWuInfo::getWorkunitThorSlaveLogFromZAPFile(const char* thorProcess, const char* logDate, int slaveNum,
+    MemoryBuffer& buf, const char* outFile)
+{
+    StringBuffer zapWUFolder;
```
should avoid referring to as 'zap'

wangkx replied:
Will do

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.hpp:
```diff
@@ -147,12 +149,19 @@ class WsWuInfo
+    {
+        Owned<IStringIterator> debugValues(&cw->getDebugValues("imported"));
+        if (debugValues->first())
+            debugValues->str(debugImportedStr);
```
I think instead do:
```
   SCMStringBuffer imported;
   cw->getDebugValue(imported);
   logsPerWorkUnit = !isEmptyString(imported.str());
```

wangkx replied:
Not sure it should be called 'logsPerWorkUnit' because the debugImportedStr is also used for # of thor slaves and log date.

jakesmith replied:
Where?
As far as I can see, it is only tested to see if set (not it's content), e.g.:
```
if (debugImportedStr.length()) //WU imported from a ZAP report
```

So is essentially a simple flag

wangkx replied:
It is a simple flag. How bout changing the debugImportedStr to hasDebugImportedStr? The flag is not only used for getting logs, also used for finding numberOfSlaves for standard WU and for Imported WU.

jakesmith commented:
@wangkx - added some more comments.

Because per workunits will eventually be the norm.
It would be best to common up as much as possible and remove references where possible of 'ZAP' and 'imported', and instead distinguish between the 2 flavours only by whether logs per workunit, or filtered logging.

Need to check everything works as expected when the ZAP file is imported and has no relation to the current environment, i.e. the thor process, group, cluster do not not exist.
So the code in ws_workunitsHelper must where possible not call into the environment to lookup info (like # of slaves), and instead rely on info in the workunit.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5389,15 +5556,44 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
+            {
+                VStringBuffer xpath("/GraphProgress/%s", wuidUnique);
+                Owned<IRemoteConnection> progressConn = querySDS().connect(xpath, myProcessSession(), RTM_LOCK_WRITE|RTM_CREATE_QUERY, SDS_LOCK_TIMEOUT);
+                if (progressConn)
```
Will throw an exception

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5389,15 +5556,44 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
+                {
+                    IPropertyTree *progressRoot = progressConn->queryRoot();
+                    if (progressRoot->hasChildren())
+                        WARNLOG("_createWorkUnit WUID %s graphprogress already exists, replacing", wuidUnique);
```
Will remove the check.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5389,15 +5556,44 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
+            }
+
+            if (root->hasChildren())
+                WARNLOG("_createWorkUnit WUID %s replacing wuXML", wuidUnique);
```
Will remove the check.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5389,15 +5556,44 @@ class CDaliWorkUnitFactory : public CWorkUnitFactory, implements IDaliClientShut
+
+            IPropertyTree *root = conn->queryRoot();
+            const char *wuidUnique = root->queryName();
+            wuXMLPTree->renameProp("/", wuidUnique);
```
ok

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5062,171 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+            getConfigurationDirectory(directories, "data", component, instance, unzipDir);
+            unzipDir.append(PATHSEPSTR).append(zapReportFileName.length() - 4, zapReportFileName);
+            if (!user.isEmpty())
+                unzipDir.append("_").append(user);
```
Will do

HPCCSmoketest commented:
Process of PR-13111, label: h22136 starts now.
The reason of this test is: New commit.
Commit ID: 7ac49633cbf2834e00b3de5f2bcd671d27d9cff6
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 7ac49633cbf2834e00b3de5f2bcd671d27d9cff6
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |38 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |865 |865 |0 |174 sec  (00:02:54) |
|test (thor) |791 |791 |0 |723 sec  (00:12:03) |
|test (roxie) |942 |942 |0 |419 sec  (00:06:59) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   42 sec (00:00:42) |  262 sec (00:04:22) |    0 sec (00:00:00) |    2 sec (00:00:02) |   20 sec (00:00:20) | 1634 sec (00:27:14) |   21 sec (00:00:21) | 1981 sec (00:33:01) |

HPCCSmoketest commented:
Process of PR-13111, label: h22136 starts now.
The reason of this test is: New commit.
Commit ID: a8f3bc3aba3e741fd1a614d02d9b5501621da18a
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: a8f3bc3aba3e741fd1a614d02d9b5501621da18a
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |21 sec  (00:00:21) |
|test (hthor) |865 |865 |0 |205 sec  (00:03:25) |
|test (thor) |791 |791 |0 |715 sec  (00:11:55) |
|test (roxie) |942 |942 |0 |254 sec  (00:04:14) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   31 sec (00:00:31) |  260 sec (00:04:20) |    0 sec (00:00:00) |    3 sec (00:00:03) |   21 sec (00:00:21) | 1513 sec (00:25:13) |   22 sec (00:00:22) | 1850 sec (00:30:50) |

wangkx commented:
@jakesmith please review my changes and let me know if I miss anything.

jakesmith commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1084,7 +1140,9 @@ unsigned WsWuInfo::getWorkunitThorLogInfo(IArrayOf<IEspECLHelpFile>& helpers, IE
 
             StringBuffer groupName;
-            getClusterThorGroupName(groupName, processName.str());
+            cw->getThorGroup(groupName);
```
see other comment.
I know I suggested it, but as you can see now you've implemented, it doesn't make sense, because the group can be different depending on 'processName', i.e. Thor can run on >1 group.
Need to remove getThorGroup/setThorGroup.

wangkx replied:
Will remove

wangkx replied:
Can one WU use >1 thor group?

jakesmith replied:
yes.
A workunit can run on any of the Thor's attached to the queue, and there can be any number of Thors, which may be on same group, or independent groups.

jakesmith replied:
in fact I was wrong, all Thor's in a logical cluster must be the same size *at the moment*.
It would be best to try not to assume that's true though, as it may well change.

jakesmith commented:
@wangkx - haven't been through it all, but it doesn't feel encapsulated enough yet. i.e. there's too much conditional code depending on where the workunit came from.
Ideally the esp code would call methods in the workunit to get all the necessary info., such that esp code dealing with workunits regardless of whether they're imported or not, was the same.
Imported workunits can have extra info. added, to help the workunit methods return the correct info.

Standard workunits can do what they used to do, like lookup clusters in the environment, lookup groups etc., but it would be good to try to get the esp code not to manually do this, it should be encapsulated in workunit.* as much as possible.

wangkx commented on esp/bindings/http/platform/httptransport.cpp:
```diff
@@ -1957,19 +1957,18 @@ bool CHttpRequest::readUploadFileName(CMimeMultiPart* mimemultipart, StringBuffe
-    tmpFileName.appendf("%s/%s.part", filePath, fileName.str());
+    StringBuffer tmpFileName(filePath);
+    if (tmpFileName.charAt(tmpFileName.length() - 1) != PATHSEPCHAR)
+        tmpFileName.append(PATHSEPCHAR);
```
Will do

wangkx commented on esp/bindings/http/platform/httptransport.cpp:
```diff
@@ -1957,19 +1957,18 @@ bool CHttpRequest::readUploadFileName(CMimeMultiPart* mimemultipart, StringBuffe
+    if (tmpFileName.charAt(tmpFileName.length() - 1) != PATHSEPCHAR)
+        tmpFileName.append(PATHSEPCHAR);
+    tmpFileName.append(pathTail(fileName));
+    fileNameWithPath.set(tmpFileName);
```
OK

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
+    ForEach(*thorSlaveFiles)
+    {
+        const char* pStr = strstr(thorSlaveFiles->query().queryFilename(), "_thorslave.");
+        pStr += 11;
```
Will change it to strlen.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
+        slaveNumStr.append(slaveNumStrEnd - pStr, pStr);
+        unsigned slaveNum = atoi(slaveNumStr);
+        if (slaveNum > numberOfSlaves)
+            numberOfSlaves = slaveNum;
```
The format of the slave log in ZAP report is not n_of_m. For example, this is a slave log: thor02_160way_thorslave.9.2019_11_07

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
+        pStr += 11;
+        const char* slaveNumStrEnd = strchr(pStr, '.');
+        if (!slaveNumStrEnd)
+            continue;
```
In the ZAP report, the name of the slave log file does not contain 'of'.

wangkx committed:
HPCC-22136 Change more based on review

Remove getThorGroup()/setThorGroup();
Add getSlaveLogPattern()/setSlaveLogPattern();
Add readSlaveLogPattern();
Add logSingleFile();
Add getNumberOfThorSlaves();
Add getWUThorLogInfo();
Switch to addPathSepChar();
Change hardcoded constants to strlen;
Switch the debugImportedStr to logSingleFile.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-13111, label: h22136 starts now.
The reason of this test is: New commit.
Commit ID: c243213841fc80de50389fe031020137ae34bf69
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: c243213841fc80de50389fe031020137ae34bf69
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |36 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |29 sec  (00:00:29) |
|test (hthor) |866 |866 |0 |183 sec  (00:03:03) |
|test (thor) |801 |801 |0 |735 sec  (00:12:15) |
|test (roxie) |943 |943 |0 |236 sec  (00:03:56) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  259 sec (00:04:19) |    0 sec (00:00:00) |    3 sec (00:00:03) |   20 sec (00:00:20) | 1481 sec (00:24:41) |   21 sec (00:00:21) | 1818 sec (00:30:18) |

HPCCSmoketest commented:
Process of PR-13111, label: h22136 starts now.
The reason of this test is: New commit.
Commit ID: 6b40c494ba648241e31158f5e8f06a31a0696571
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 6b40c494ba648241e31158f5e8f06a31a0696571
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |47 sec  (00:00:47) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |866 |866 |0 |173 sec  (00:02:53) |
|test (thor) |801 |801 |0 |728 sec  (00:12:08) |
|test (roxie) |943 |943 |0 |238 sec  (00:03:58) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   34 sec (00:00:34) |  256 sec (00:04:16) |    0 sec (00:00:00) |    3 sec (00:00:03) |   20 sec (00:00:20) | 1452 sec (00:24:12) |   22 sec (00:00:22) | 1787 sec (00:29:47) |

HPCCSmoketest commented:
Process of PR-13111, label: h22136 starts now.
The reason of this test is: New commit.
Commit ID: 2e9f2f05bdb233410f01dd950042c78f284e7fbb
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: 2e9f2f05bdb233410f01dd950042c78f284e7fbb
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |37 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |3 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |45 sec  (00:00:45) |
|setup (roxie) |11 |11 |0 |24 sec  (00:00:24) |
|test (hthor) |866 |866 |0 |205 sec  (00:03:25) |
|test (thor) |801 |801 |0 |736 sec  (00:12:16) |
|test (roxie) |943 |943 |0 |225 sec  (00:03:45) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   33 sec (00:00:33) |  259 sec (00:04:19) |    0 sec (00:00:00) |    3 sec (00:00:03) |   20 sec (00:00:20) | 1488 sec (00:24:48) |   22 sec (00:00:22) | 1825 sec (00:30:25) |

wangkx commented:
@jakesmith please review my changes.

HPCCSmoketest commented:
Process of PR-13111, label: h22136 starts now.
The reason of this test is: New commit.
Commit ID: d9d53a373a013d756dcabcad2b0048521e6b0880
Estimated completion time is ~0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-327.28.3.el7.x86_64)
 GCC:gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Sha: d9d53a373a013d756dcabcad2b0048521e6b0880
Build: success
Build: success
Install HPCC Platform
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |116 |116 |0 |0 |0 |35 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |26 sec  (00:00:26) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |866 |866 |0 |186 sec  (00:03:06) |
|test (thor) |801 |801 |0 |730 sec  (00:12:10) |
|test (roxie) |943 |943 |0 |422 sec  (00:07:02) |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   33 sec (00:00:33) |  262 sec (00:04:22) |    0 sec (00:00:00) |    2 sec (00:00:02) |   20 sec (00:00:20) | 1678 sec (00:27:58) |   21 sec (00:00:21) | 2016 sec (00:33:36) |

jakesmith commented on plugins/cassandra/cassandrawu.cpp:
```diff
@@ -3259,6 +3374,7 @@ class CCasssandraWorkUnitFactory : public CWorkUnitFactory, implements ICassandr
                 wuXML->setProp("@xmlns:xsi", "http://www.w3.org/1999/XMLSchema-instance");
                 wuXML->setPropInt("@wuidVersion", WUID_VERSION);  // we implement the latest version.
                 wuXML->setProp("@totalThorTime", ""); // must be non null, otherwise sorting by thor time excludes the values
+
```
trivial: unintentional extra line added?

wangkx replied:
Will remove

jakesmith commented:
@wangkx - looks better, I think getting closer, have added some new comments.
May be worth discussing some of the points in flowdock to clarify.

wangkx commented on esp/services/ws_workunits/ws_workunitsService.cpp:
```diff
@@ -4442,6 +4444,46 @@ int CWsWorkunitsSoapBindingEx::onGetForm(IEspContext &context, CHttpRequest* req
+
+            request->readContentToFiles(nullptr, zipFolder, fileNames);
+            if (!fileNames.ordinality())
+                throw MakeStringException(-1, "Failed to read upload content.");
```
Will loop through each file name and import > 1 WUs if available.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -13778,3 +14153,23 @@ bool isValidMemoryValue(const char *memoryUnit)
+    Owned<IConstEnvironment> env = envFactory->openEnvironment();
+    Owned<IPropertyTree> root = &env->getPTree();
+
+    Owned<IWorkUnitFactory> wuFactory = getWorkUnitFactory();
```
I may call the getWorkUnitFactory() when initializing WsWorkunits service and call the wuFactory->importWorkUnit(...) directly. Then, I may remove this WORKUNIT_API: importWorkunitFromZAPFile().

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -13778,3 +14153,23 @@ bool isValidMemoryValue(const char *memoryUnit)
+    if (isEmptyString(component))
+        throw MakeStringException(WUERR_InvalidUserInput, "Empty Component name");
+    if (isEmptyString(instance))
+        throw MakeStringException(WUERR_InvalidUserInput, "Empty Instance name");
```
Will change it to makeStringException().

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5087,174 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+        void buildPTreesFromZAPReport()
+        {
+            readWUXMLFileToPTree();
+            wuTree->renameProp("/", wuid.get()); //renameProp(nullptr, wuid.get()) not work: renameProp: cannot rename self, renameProp has to rename in context of a parent
```
Will remove.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5087,174 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+        {
+            readWUXMLFileToPTree();
+            wuTree->renameProp("/", wuid.get()); //renameProp(nullptr, wuid.get()) not work: renameProp: cannot rename self, renameProp has to rename in context of a parent
+            setImportDebugAttribute();
```
Ok, will call the existing API.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5087,174 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+            str.append(" at ").append(unzipDateTime);
+            str.append(" from ").append(zapReportFileName);
+            wuTree->setProp("Debug", "");
+            wuTree->setProp("Debug/imported", str);
```
The code for unzipDateTime may be removed by a mistake. I will add it back.
The scope stores who originally created the WU in the ZAP report. The intention for adding user is to record who imports the WU.  @jakesmith do you think there is no need to record who imports the WU?

jakesmith replied:
>The scope stores who originally created the WU in the ZAP report.

I'd have to check, but I thought what you were passing through to create thew new workunit was the current user, so same as user importing, and therefore original owner was lost?
If that's so, then no need to store separately in Debug/@imported.

wangkx replied:
OK, I may change the scope in the old WU XML PTree to be the userid of the importing user before importing it.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -6200,6 +6402,40 @@ class CStringPTreeAttrIterator : implements IStringIterator, public CInterface
+
+//==========================================================================================
+
+class CWUThorLogInfo: public CSimpleInterface, implements IConstWUThorLogInfo
```
Will change it.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -10013,6 +10249,145 @@ IWUGraphStats *CLocalWorkUnit::updateStats(const char *graphName, StatisticCreat
+            }
+
+            datePtr += thorMasterLogDateSearchString.length();
+            if (strlen(datePtr) < 10) //log single file
```
Will add a comment

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -10013,6 +10249,145 @@ IWUGraphStats *CLocalWorkUnit::updateStats(const char *graphName, StatisticCreat
+            logDate.append(10, datePtr);
+
+            Owned<CWUThorLogInfo> thorLog = new CWUThorLogInfo(logName.str(), processName.str(), groupName, logDate);
+            wuThorLogs.append(*thorLog.getLink());
```
Will change

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -10013,6 +10249,145 @@ IWUGraphStats *CLocalWorkUnit::updateStats(const char *graphName, StatisticCreat
+    if (hasDebugValue("imported")) //WU imported from a ZAP report
+        thorMasterLogDateSearchString.set("_thormaster.");
+    else
+        thorMasterLogDateSearchString.set("/thormaster.");
```
How about adding 'thorMasterLogSearchString=_thormaster.' in Debug/imported?

wangkx replied:
I am changing the WsWorkunits code to depr_ver the log date from Thor Log Info.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5087,174 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+            graphProgressTree.setown(createPTreeFromXMLFile(fileName));
+            if (!graphProgressTree)
+                throw MakeStringException(WUERR_InvalidUserInput, "Failed to retrieving ZAPReport_*.graphprogress.");
+        }
```
Will change.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5087,174 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+        }
+        void setUNZIPDir(const IPropertyTree *directories, const char *component, const char *instance)
+        {   //Set a unique unzip folder inside the component's data folder
+            getConfigurationDirectory(directories, "data", component, instance, unzipDir);
```
Will change similar to getWorkUnitFactory().

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5087,174 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+    };
+
+    CImportWorkUnitHelper helper(zapReportFileName, user);
+    Owned<IWorkUnit> newWU = createWorkUnit(app, user, secMgr, secUser);
```
Please see my other comment.

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -10013,6 +10249,145 @@ IWUGraphStats *CLocalWorkUnit::updateStats(const char *graphName, StatisticCreat
+    unsigned numberOfThorSlaves = 0;
+    StringBuffer slaveLogPattern;
+    getSlaveLogPattern(slaveLogPattern);
+    if (!slaveLogPattern.isEmpty())
```
Yes, the getNumberOfThorSlaves is all I need for now. I will modify addProcess() to add 'slaves=n'.

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2051,6 +2061,20 @@ void WsWuInfo::getWorkunitThorSlaveLog(IPropertyTree* directories, const char *p
+    StringBuffer logPath;
+    getWorkunitProcessLogPath("Thor", logPath);
+
+    VStringBuffer logFileName("%s%c%s_thorslave.%u.%s.log", logPath.str(), PATHSEPCHAR, thorProcess, slaveNum, logDate);
```
How about adding 'thorSlaveLogSearchString=_thorslave' under Debug/imported.

jakesmith replied:
Relating to comments elsewhere, my feeling is that this would be cleaner if all pertanent info. was captured in the workunit under Process/ including whether @singleLog.

You could then expect IConstWorkUnit::getLogs() to retrieve any log, including slave logs.
But you might want to extend it, so it can give you a specific slave log, e.g:
```
IStringIterator *CLocalWorkUnit::getLogs(const char *type, const char *instance, unsigned n) const
{
     // See if Process/<type>/<instance> has a @pattern .. If has @pattern, should also have @max (for # slaves)
     // if has pattern and n>0, construct log using pattern and n. Error if n>max.
     // if has pattern and n undefined/default, construct logs for all up to @max.
}
```

```
Owned<IStringIterator> logs = workunit->getLogs("Thor", "mythor", 5);
```

jakesmith replied:
Also, the introduction of IConstWUThorLogInfo may be sensible, but it's questionable why there is now 2 different ways to get log info getWUThorLogInfo() and getLogs()

I suspect you can simplify, so that everything uses getWUThorLogInfo ?

jakesmith replied:
And I think when simplified, your code WsWuInfo::getWorkunitThorSlaveLogSingleFile code may go away.
With everything universally using IConstWUThorLogInfo (which will encode whether @snigle or not too), then your code based on IConstWUThorLogInfo::isSingleLog(), calling your getWorkunitLogSingleFile

wangkx replied:
Will check

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.cpp:
```diff
@@ -2985,6 +3009,7 @@ bool addToQueryString(StringBuffer &queryString, const char *name, const char *v
 int WUSchedule::run()
 {
     PROGLOG("ECLWorkunit WUSchedule Thread started.");
+
```
Will remove

wangkx commented on esp/services/ws_workunits/ws_workunitsHelpers.hpp:
```diff
@@ -137,6 +137,8 @@ struct WsWUExceptions
 
 class WsWuInfo
 {
+    bool logSingleFile = false;
```
Will change

wangkx commented on common/workunit/workunit.hpp:
```diff
@@ -1286,6 +1295,10 @@ interface IConstWorkUnit : extends IConstWorkUnitInfo
     virtual IStringVal & getAbortBy(IStringVal & str) const = 0;
     virtual unsigned __int64 getAbortTimeStamp() const = 0;
+    virtual StringBuffer & getSlaveLogPattern(StringBuffer & str) const = 0;
+    virtual bool logSingleFile() const = 0;
```
Will change

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -4041,6 +4056,18 @@ class CLockedWorkUnit : implements ILocalWorkUnit, implements IExtendedWUInterfa
+    virtual StringBuffer &getSlaveLogPattern(StringBuffer &str) const
+            { return c->getSlaveLogPattern(str); }
+    virtual void setSlaveLogPattern(const char *pattern)
+            { return c->setSlaveLogPattern(pattern); }
```
No need for setSlaveLogPattern/getSlaveLogPattern. Will remove.

wangkx replied:
@jakesmith I am adding the code to set a 'pattern for slave log' when importing a WU. Should I also add some code to set the 'pattern for slave log' when a standard thor WU is created? The existing thgraphmanager.cpp calls addProcess() to create the /Progress/Thor section in the CJobManager::updateWorkUnitLog(). I am not sure that is the place to set the 'pattern for slave log' (no enough information about slave log?). Any suggestion?

jakesmith replied:
Yes, see other comments, I think if you capture enough info (including #slaves in Thor case) via addProcess, then things become a lot cleaner.
Thor knows the pattern of the slave log (which will change at some point), so yes updateWorkUnitLog is a reasonable starting point.

I think I'd add @max (for #slaves), and @pattern for when retrieving/building slave logs, e.g. with:
```
void addProcess(const char *type, const char *instance, unsigned pid, const char *log, unsigned max, const char *pattern)
```

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -10013,6 +10249,145 @@ IWUGraphStats *CLocalWorkUnit::updateStats(const char *graphName, StatisticCreat
+            continue;
+
+        bool *found = uniqueProcesses.getValue(processName.str());
+        if (found && *found)
```
Ok

wangkx commented on common/workunit/workunit.cpp:
```diff
@@ -5060,6 +5087,174 @@ bool CWorkUnitFactory::restoreWorkUnit(const char *base, const char *wuid, bool
+    helper.setUNZIPDir(directories, component, instance);
+    helper.unzipZAPReport(zapReportFilePath, zapReportPassword);
+    helper.buildPTreesFromZAPReport();
+    newWU->import(helper.queryWUPTree(), helper.querygraphProgressPTree());
```
Should I also remove the ZAP report file (the zip file being uploaded)?

jakesmith replied:
yes I think should

wangkx committed:
HPCC-22136 Revise based on review - WIP

Made many changes and still have some TODOs.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

jakesmith commented:
@wangkx - have added more comments.

There is still probably some significant changes to make this simpler and more encapsulated.
This PR is getting a bit bloated with comments etc.

Can you look at latest comments and close this PR and open a new one when ready?

wangkx commented:
I am closing this PR. A new PR will be opened after more changes to make this simpler and more encapsulated.
